extension/fluentbitextension/process.go

```diff
@@ -0,0 +1,207 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fluentbitextension
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+type processManager struct {
+	cancel context.CancelFunc
+	conf   *Config
+	logger *zap.Logger
+}
+
+func newProcessManager(conf *Config, logger *zap.Logger) *processManager {
+	return &processManager{
+		conf:   conf,
+		logger: logger,
+	}
+}
+
+type procState string
+
+// A global var that is available only for testing
+var restartDelay = 10 * time.Second
+
+const (
+	Uninitialized procState = "uninitialized"
+	Running       procState = "running"
+	Starting      procState = "starting"
+	ShuttingDown  procState = "shutting-down"
+	Stopped       procState = "stopped"
+	Restarting    procState = "restarting"
+	Errored       procState = "errored"
+)
+
+func constructArgs(tcpEndpoint string) []string {
+	return []string{
+		"--config=/dev/stdin",
+		"--http",
+		"--port=2020",
+		"--flush=1",
+		"-o", "forward://" + tcpEndpoint,
+		"--match=*",
+	}
+}
+
+func (pm *processManager) Start(ctx context.Context, host component.Host) error {
+	childCtx, cancel := context.WithCancel(ctx)
+	pm.cancel = cancel
+
+	args := pm.conf.Args
+	if len(args) == 0 {
+		args = constructArgs(pm.conf.TCPEndpoint)
+	}
+	go run(childCtx, pm.conf.ExecutablePath, args, pm.conf.Config, pm.logger)
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (pm *processManager) Shutdown(ctx context.Context) error {
+	pm.cancel()
+	return nil
+}
+
+func run(ctx context.Context, execPath string, args []string, config string, logger *zap.Logger) {
+	state := Uninitialized
+
+	var cmd *exec.Cmd
+	var err error
+	var stdin io.WriteCloser
+	var stdout io.ReadCloser
+	// procWait is guaranteed to be sent exactly one message per successful process start
+	procWait := make(chan error)
+
+	// A state machine makes the management easier to understand and account
+	// for all of the edge cases when managing a subprocess.
+	for {
+		logger.Debug("Fluent extension changed state", zap.String("state", string(state)))
+
+		switch state {
+		case Errored:
+			logger.Error("FluentBit process died", zap.Error(err))
+			state = Restarting
+
+		case Uninitialized:
+			logger.Debug("Starting fluent subprocess", zap.String("execPath", execPath))
+			cmd, stdin, stdout = createCommand(execPath, args)
+			state = Starting
+		case Starting:
```

Comment:
Uninitialized and Starting appear to be logically one state, the separation looks artificial. I can't see a jump to Starting that bypasses Uninitialized.

Reply:
You're right, they are (was a leftover from the code I copied in Smart Agent where the distinction was more real).  Removed the Starting state.